### PR TITLE
Add GeneralSoftGrudger Strategy

### PR DESCRIPTION
**Pull request by jtsmith2**
A generalization of the SoftGrudger strategy. SoftGrudger punishes by playing: D, D, D, D, C, C. after a defection by the opponent.  GeneralSoftGrudger only punishes after its opponent defects a specified amount of times consecutively (parameter "n"). The punishment is in the form of a series of defections (parameter "d") followed by a 'penance' of a series of consecutive cooperations (parameter "c").

SoftGrudger (already existed) is "soft spiteful" from the PRISON list in #379 and this is a generalization of that strategy.

### CODE REVIEW

File: axelrod/strategies/grudger.py [hunk 1]
```diff
@@ -239,3 +239,73 @@ def strategy(opponent: Player) -> Action:
...
         return D
+
+class GeneralSoftGrudger(Player):
+    """
+    A generalization of the SoftGrudger strategy. SoftGrudger punishes by playing:
+    D, D, D, D, C, C. after a defection by the opponent.  GeneralSoftGrudger
+    only punishes after its opponent defects a specified amount of times consecutively.
+    The punishment is in the form of a series of defections followed by a 'penance' of
+    a series of consecutive cooperations.
+    """
```
Comment (drvinceknight) [inline axelrod/strategies/grudger.py]: Could you add:

```
Names:

- General Soft Grudger: ...
```

and after the name, either a reference to the literature (if one exists to your knowledge) **or** say that it's an original name, like here: http://axelrod.readthedocs.io/en/latest/reference/all_strategies.html#axelrod.strategies.ann.EvolvedANN

The idea of that is that it auto generates http://axelrod.readthedocs.io/en/latest/reference/all_strategies.html which is a place that users can `ctrl + F` strategies.

File: axelrod/strategies/grudger.py [hunk 2]
```diff
@@ -239,3 +239,73 @@ def strategy(opponent: Player) -> Action:
...
+        'c' consecutive cooperations.
+        """
+        if self.grudged:
+            strategy = self.grudge[self.grudge_memory]
+            self.grudge_memory += 1
+            if self.grudge_memory == len(self.grudge):
+                self.grudged = False
+                self.grudge_memory = 0
+            return strategy
+        elif [D]*self.n == opponent.history[-self.n:]:
```
Comment (drvinceknight) [inline axelrod/strategies/grudger.py]: PEP8: `[D] * self.n`

File: axelrod/strategies/grudger.py [hunk 3]
```diff
@@ -239,3 +239,73 @@ def strategy(opponent: Player) -> Action:
...
+
+        Special Cases
+        -------------
+        GeneralSoftGrudger(1,4,2) is equivalent to SoftGrudger
+        """
+        super().__init__()
+        self.n = n
+        self.d = d
+        self.c = c
+        self.grudge = [D]*(d-1)+[C]*c
```
Comment (drvinceknight) [inline axelrod/strategies/grudger.py]: PEP8: `[D] * (d - 1) + [C] * c`

**Comment by jtsmith2** [comment]
Small fixes requested by @drvinceknight should be implemented now.

**Comment by drvinceknight** [comment]
Looks great thanks!

**Pull request merged**